### ISSUE TITLE
Completely unable to build

### ISSUE THREAD

**Issue by squinkylabs** [rubyglow/racktool#1]
I tried building against 0.6.0 (mis-read the instructions). Tried building against my current 0.6. Got tons and tons of linker error, mostly having to do with the infamous openSSL.

going to get fresh pulls in rack, re-build, and try again.

**Comment by squinkylabs** [comment]
nope, did a full clean of everything and still get thousands of linker errors. Andrew might be able to advise on this.

```
-ljansson -lspeexdsp -lzip -lz -lcurl -lssl -lrtmidi -lpthread -lopengl32 -lgdi32 -lws2_32 -lcomdlg32 -lole32 -ldsound -lwinmm -lksuser -lshlwapi
D:/VCVRack/x6/build/src/util/request.cpp.o: In function `rack::requestSHA256File(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)':
D:\VCVRack\x6/src/util/request.cpp:169: undefined reference to `SHA256_Init'
D:\VCVRack\x6/src/util/request.cpp:174: undefined reference to `SHA256_Update'
D:\VCVRack\x6/src/util/request.cpp:176: undefined reference to `SHA256_Final'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.text+0x16): undefined reference to `MD5_Init'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.text+0x34): undefined reference to `MD5_Update'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.text+0x3f): undefined reference to `MD5_Final'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.rdata+0x0): undefined reference to `MD5_Init'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.rdata+0x8): undefined reference to `MD5_Update'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.rdata+0x10): undefined reference to `MD5_Final'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.rdata+0x20): undefined reference to `MD5_Init'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.rdata+0x28): undefined reference to `MD5_Update'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-md5.o):md5.c:(.rdata+0x30): undefined reference to `MD5_Final'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-sha256.o):sha256.c:(.text+0x19): undefined reference to `SHA256_Init'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-sha256.o):sha256.c:(.text+0x37): undefined reference to `SHA256_Update'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-sha256.o):sha256.c:(.text+0x42): undefined reference to `SHA256_Final'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x33): undefined reference to `EVP_MD_CTX_new'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x3b): undefined reference to `EVP_sha256'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x49): undefined reference to `EVP_DigestInit_ex'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x57): undefined reference to `EVP_DigestUpdate'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x67): undefined reference to `EVP_DigestFinal_ex'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x6f): undefined reference to `EVP_MD_CTX_free'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x97): undefined reference to `EVP_MD_CTX_new'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x9f): undefined reference to `EVP_md5'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0xad): undefined reference to `EVP_DigestInit_ex'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0xbb): undefined reference to `EVP_DigestUpdate'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0xcb): undefined reference to `EVP_DigestFinal_ex'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0xd3): undefined reference to `EVP_MD_CTX_free'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0xea): undefined reference to `ENGINE_get_first'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0xfe): undefined reference to `ENGINE_get_next'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x111): undefined reference to `ENGINE_get_id'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x154): undefined reference to `ENGINE_finish'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x160): undefined reference to `ENGINE_free'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x189): undefined reference to `ENGINE_by_id'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x1a6): undefined reference to `ENGINE_finish'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x1b2): undefined reference to `ENGINE_free'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x1c5): undefined reference to `ENGINE_init'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x1e5): undefined reference to `ENGINE_free'
D:/VCVRack/x6/dep/lib\libcurl.a(libcurl_la-openssl.o):openssl.c:(.text+0x1ea): undefined reference to `ERR_get_error'
D:/VCVRack/x6```

**Comment by rubyglow** [comment]
Thanks for trying Bruce! I think you've been caught in the fact, that the 0.6 branch has been a bit... in flux lately. But I see that 0.6.1 is just out, and the branch should have settled down and has been tagged. I'll absorb the latest changes into my make and give you a hollar here, when it should be Ok to try again. For what it's worth, my current Rack build is at commit c6be92e5054e0aa8c4942ee7a1f0c67975fc141e

**Comment by squinkylabs** [comment]
Sounds right. tx.

**Comment by rubyglow** [comment]
Hi Bruce
I've updated the instructions at https://github.com/rubyglow/racktool/blob/master/README.md.

If you would be so kind, please start over following those, starting by "git pull" of racktool.

Basically racktool now tracks the exact released version 0.6.1 of Rack, and it's far better to build Rack in a separate directory for that purpose.

If people already have the Rack source built, at another version, it's almost impossible to explain how to *fully* clean it, *then* check out the exact version, *then* build it again, etc. etc... :-) Better just to build a seperate version for that purpose.

Remember to adjust your RACK_DIR to the new Rack-0.6.1 directory.

Thanks a bunch Bruce. Much appreciated!

**Comment by squinkylabs** [comment]
I'll probably try this weekends. And I certainly agree - it's easier to start a new repo than to mess around with an old one.

**Comment by rubyglow** [comment]
Note: I've created the loading problem itself as issue #2

**Comment by rubyglow** [comment]
The instructions for running the thing are now completely fleshed out in the README.

**Comment by squinkylabs** [comment]
the new instructions are possible less good than before?
1) they never mention `make dep`. Does this version of Rack not need dep any more?
2) they never mention cloning your repo.
3) they say to go into dep/racktool, but there is no such directly.

I'm going to guess that 
a) you do the recursive pull and make dep just like in normal rack.
b) one is supposed to go do the dep folder and clone your repo from there?

**Comment by rubyglow** [comment]
I'm sorry the instructions are not clear enough Bruce, I'll work some more on them.
So.....

The first part of the building instructions say "First build Rack", because what you are supposed to do first, is to clone and build VCV Rack (!) at the released version 0.6.1. It links to the README of Rack, because you are supposed to build a version of Rack, like Andrew says there, but not at HEAD of his 0.6 branch, but rather at the v0.6.1 tag. It ends by saying that you should verify that VCV Rack build, by doing a make run on it. So that whole first part is *only* about building VCV Rack.

You must now set the RACK_DIR variable pointing to the Rack-0.6.1 directory that you cloned before, so e.g. `set RACK_DIR=C:\path\to\Rack-0.6.1`. This is because rackproxy *depends* on the build of VCV Rack, which you have now just made.

The second part of the build instructions start with "Then set the RACK_DIR", and yes, that presumes that now you've done a git clone of this "racktool" repository, with: git clone https://github.com/rubyglow/racktool.git
I will make sure to spell that out. So when you've cloned my racktool repository you now have the dep/rackproxy directory. You now go into that and run make, so e.g:
cd /path/to/racktool/dep/rackproxy

Now the "racktool" program has been built.

So, to summarize - you are git cloning two independant repositories; VCV Rack and racktool.
You then first build VCV Rack, at a specific version.
You then secondly build a sub-program of racktool, called rackproxy, which depends on the Rack build during it's own building.

Are there still parts of this you find unclear?
Thanks Bruce.

**Comment by rubyglow** [comment]
The following is the exact set of commands, that you could literally paste into the MinGW 64-bit terminal on Windows, to build Rack and then rackproxy on Windows. It presumes of course, that you have already installed and tested the MSYS2 based toolchain on you Windows machine, by sucessfully completing the "Setting up your development environment" section in the VCV Rack README.

In this example I'm using "make -j2" to build using two CPU cores. If you have 4 cores you can substitute with -j4, etc.

\# BEGIN instructions

\# Create a base source directory on the machine
\# The base directory is c:\src\rackdev\ in Windows land
mkdir -p /c/src/rackdev/

\# Clone and build VCV Rack 0.6.1
cd /c/src/rackdev/
git clone -b v0.6.1 https://github.com/VCVRack/Rack.git Rack-0.6.1
cd Rack-0.6.1
git submodule update --init --recursive
make -j2 dep
make -j2

\# Now verify that the VCV Rack build is succesful, by running it, then close it afterwards.
make run

\# We're done with Rack, now clone and build racktool/rackproxy
export RACK_DIR=/c/src/rackdev/Rack-0.6.1
cd /c/src/rackdev/
git clone https://github.com/rubyglow/racktool.git
cd racktool/dep/rackproxy
make -j2

\# END instructions

Now rackproxy has been built. In Windows land it is located in c:\src\rackdev\racktool\dep\rackproxy\.

Time to test it in Windows, like any other Windows user would.
In the following example I'll presume that your MSYS2 installation is at c:\msys\ in Windows. Substitute that if you installed it somewhere else.

Open a regular Windows console/command-line/command-prompt (cmd.exe)
You can literally copy and paste the following examples into the console.

@echo off

rem Go to where rackproxy lives
c:
cd c:\src\rackdev\racktool\dep\rackproxy\

rem Include needed libraries from MinGW in your PATH
set PATH=c:\msys\mingw64\bin;%PATH%

rem Show usage and verify the program even runs
rackproxy help

rem Display metadata about the Core plugin as JSON
rackproxy plugin core

rem Display metadata about the Fundamental plugin as JSON
rem Here is where things break down with a code 126 error
rem This example presumes you already have a Rack 0.6.* installation on this machine.
rem If not, simply download some other Rack 0.6.* plugin for Windows, and unzip it somewhere.
rem You could use e.g. https://github.com/cfoulc/cf/releases/download/0.6.0/cf-0.6.0-win.zip
rackproxy plugin "%USERPROFILE%\Documents\Rack\plugins\Fundamental"

**Comment by squinkylabs** [comment]
thanks! The one part I couldn't be sure to guess correctly was where to clone racktools. I'll try it out.

**Comment by squinkylabs** [comment]
built and ran fine! Now to investigate error 126. btw: I ran it from the msys2 shell and did not have to set any env variables to run. 
`$ ./rackproxy.exe plugin /d/VCVRack/6rack/plugins/Fundamental/`

**Comment by squinkylabs** [comment]
Ok, I got it to run without the error. I just had to put dep/lib in my path:
`set PATH=c:\msys64\mingw64\bin;d:\VCVRack\rackdev\Rack-0.6.1;d:\VCVRack\rackdev\Rack-0.6.1\dep\lib;%PATH%`

Although it didn't turn out to matter, you are passing LoadLibrary a string with mixed back and forward slashes. LoadLibrary docs claim the must be backslashes. When you run from msys2 instead of cmd it's all forward slashes.

Here's the output I got for Fundamental:
```
{"dir": "d:\\VCVRack\\6rack\\plugins\\Fundamental", "file": "d:\\VCVRack\\6rack\
\plugins\\Fundamental/plugin.dll", "loaded": true, "loadError": "", "slug": "Fun
damental", "version": "0.6.0", "numModules": 16, "modules": [{"slug": "VCO", "na
me": "VCO-1", "author": "Fundamental", "tags": ["Oscillator/VCO"]}, {"slug": "VC
O2", "name": "VCO-2", "author": "Fundamental", "tags": ["Oscillator/VCO"]}, {"sl
ug": "VCF", "name": "VCF", "author": "Fundamental", "tags": ["Filter/VCF"]}, {"s
lug": "VCA", "name": "VCA", "author": "Fundamental", "tags": ["Amplifier/VCA", "
Dual/Stereo"]}, {"slug": "LFO", "name": "LFO-1", "author": "Fundamental", "tags"
: ["LFO"]}, {"slug": "LFO2", "name": "LFO-2", "author": "Fundamental", "tags": [
"LFO"]}, {"slug": "Delay", "name": "Delay", "author": "Fundamental", "tags": ["D
elay"]}, {"slug": "ADSR", "name": "ADSR", "author": "Fundamental", "tags": ["Env
elope Generator"]}, {"slug": "VCMixer", "name": "Mixer", "author": "Fundamental"
, "tags": ["Mixer", "Amplifier/VCA"]}, {"slug": "8vert", "name": "8vert", "autho
r": "Fundamental", "tags": ["Attenuator"]}, {"slug": "Unity", "name": "Unity", "
author": "Fundamental", "tags": ["Mixer", "Utility", "Dual/Stereo"]}, {"slug": "
Mutes", "name": "Mutes", "author": "Fundamental", "tags": ["Switch"]}, {"slug":
"Scope", "name": "Scope", "author": "Fundamental", "tags": ["Visual"]}, {"slug":
 "SEQ3", "name": "SEQ-3", "author": "Fundamental", "tags": ["Sequencer"]}, {"slu
g": "SequentialSwitch1", "name": "Sequential Switch 1", "author": "Fundamental",
 "tags": ["Utility"]}, {"slug": "SequentialSwitch2", "name": "Sequential Switch
2", "author": "Fundamental", "tags": ["Utility"]}]}```

**Comment by rubyglow** [comment]
Thanks! I've moved the conversation to issue #2 with a lengthy reply....

**Comment by rubyglow** [comment]
I could see that there were *way* too many details and steps that could go wrong or be misunderstood, in the building process, so I have now basically taken 90% of the pain and complications out of the process. No more futzing with RACK_DIR or manually building Rack. Rack is now fetched and built automatically as a git sub-module, and the entire build proces, everything included, has now been reduced to this:

```
git clone https://github.com/rubyglow/racktool.git
cd racktool
git submodule update --init --recursive
cd dep/rackproxy
make -j dep
make -j
```
And to run a quick verification test: `make test`
That's it :-)
If you feel like trying it, start over in a clean directory.
The README should also be a lot simpler and much more concise now, and I welcome your feedback.
Thanks for your help so far Bruce!

**Comment by squinkylabs** [comment]
I will try it, thanks. And thanks for putting the code up in it's current state ;-)
